arch/rv32i/src/lib.rs

```diff
@@ -388,15 +388,15 @@ global_asm!(
 
 /// RISC-V semihosting needs three exact instructions in uncompressed form.
 ///
-/// See https://github.com/riscv/riscv-semihosting-spec/blob/main/riscv-semihosting-spec.adoc#11-semihosting-trap-instruction-sequence
-/// for more details on the three insturctions.
+/// See <https://github.com/riscv/riscv-semihosting-spec/blob/main/riscv-semihosting-spec.adoc#11-semihosting-trap-instruction-sequence>
+/// for more details on the three instructions.
 ///
 /// In order to work with semihosting we include the assembly here
 /// where we are able to disable compressed instruction support. This
 /// follows the example used in the Linux kernel:
-/// https://elixir.bootlin.com/linux/v5.12.10/source/arch/riscv/include/asm/jump_label.h#L21
+/// <https://elixir.bootlin.com/linux/v5.12.10/source/arch/riscv/include/asm/jump_label.h#L21>
 /// as suggested by the RISC-V developers:
-/// https://groups.google.com/a/groups.riscv.org/g/isa-dev/c/XKkYacERM04/m/CdpOcqtRAgAJ
+/// <https://groups.google.com/a/groups.riscv.org/g/isa-dev/c/XKkYacERM04/m/CdpOcqtRAgAJ>
 #[cfg(all(target_arch = "riscv32", target_os = "none"))]
```

Comment:
Ideally we'd do `#[cfg(any(doc, all(target_arch = "riscv32", target_os = "none")))]` here, and then `!doc` for the other branch. This ensures that we generate the correct docs regardless of `--target` passed to Rustdoc.

Should probably be a repo-wide, distinct PR though.

Reply:
Please no more extra nonsense to fix a nonsensical problem. There is absolutely no purpose to compile a crate called "rv32i" on any architecture other than rv32i. That cargo doesn't understand that is frustrating but unless something is broken I want to avoid adding more workarounds.

Reply:
> Please no more extra nonsense to fix a nonsensical problem.

I understand the frustration, but this is a real (albeit arguably nonsensical) problem. #4054 will fix this, but the way that we call out to Rustdoc right now breaks exactly this -- https://docs.tockos.org/rv32i/fn._start_trap, for example, completely lacks docs that are present in source:

https://github.com/tock/tock/blob/9d1f488899445fd2695429ebb46409cece54dd5d/arch/rv32i/src/lib.rs#L185-L271

This is the intended way of doing this(TM): https://doc.rust-lang.org/rustdoc/advanced-features.html#cfgdoc-documenting-platform-specific-or-feature-specific-information

I don't like this, and neither do I _want_ to make this repo-wide change. I'm merely suggesting what "solution" upstream Rust has to offer.

Reply:
> There is absolutely no purpose to compile a crate called "rv32i" on any architecture other than rv32i.

For what its worth, at this point why don't we just get rid off all conditional compilation and break compilations for all other targets deliberately?

```rust
#[cfg(not(all(target_arch = "riscv32", target_os = "none")))]
compile_error!("Crate rv32i can only be build on a rv32i*-unknown-none-elf toolchain!");
```

Reply:
Wait. We can do that now? Talk about buying the lede! Hmm...suddenly hidden files don't seem so bad. (All the other problems are still there, however.)

But let's just delete the code we never would have written if `make check` just did the right thing.

Reply:
Yeah, this doesn't seem to work :(

Reply:
Yeah, it does not force cargo to build for a specific target. As discussed on the call today, we still want to support doc builds for other targets, as that's how our doc builds work (for better or for worse).

From what we talked about today, it seemed like `#[cfg(doc)]` was a somewhat reasonable option and could allow us to get rid of the `make alldocs` bodge, which seems brittle anyways. However let's table that until we have #4054 in at least.